core/domain/exp_services.py

```diff
@@ -1912,3 +1921,101 @@ def discard_draft(exp_id, user_id):
         exp_user_data.draft_change_list_last_updated = None
         exp_user_data.draft_change_list_exp_version = None
         exp_user_data.put()
+
+
+def get_state_id_mapping(exp_id, exp_version):
+    """Retrieve state id mapping model instance from the datastore.
+
+    Args:
+        exp_id: str. The exploration id.
+        exp_version: int. The exploration version.
+
+    Returnes:
+        StateIDMapping. Domain object for state id mapping model instance.
+    """
+    model = exp_models.StateIDMappingModel.get_state_id_mapping_model(
+        exp_id, exp_version)
+    state_id_mapping = exp_domain.StateIDMapping(
+        model.exploration_id, model.exploration_version,
+        copy.deepcopy(model.state_names_to_ids), model.largest_state_id_used)
+    return state_id_mapping
+
+
+def _save_state_id_mapping(state_id_mapping):
+    """Stores state id mapping instance in datastore.
+
+    Args:
+        state_id_mapping: StateIDMapping. State ID mapping which is to be
+            stored in database.
+    """
+    exp_models.StateIDMappingModel.create(
+        state_id_mapping.exploration_id,
+        state_id_mapping.exploration_version,
+        state_id_mapping.state_names_to_ids,
+        state_id_mapping.largest_state_id_used)
+
+
+def create_and_save_state_id_mapping_model(exploration, change_list):
+    """Create and store state id mapping for new exploration.
+
+    Args:
+        exploration: Exploration. Exploration for which state id mapping is
+            to be stored.
+        change_list: list(dict). A list of changes made in the exploration.
+
+    Returns:
+        StateIDMapping. Domain object of StateIDMappingModel instance.
+    """
+    if exploration.version > 1:
+        # Get state id mapping for new exploration from old exploration with
+        # the help of change list.
+        old_exploration = get_exploration_by_id(
+            exploration.id, version=exploration.version - 1)
+        old_state_id_mapping = get_state_id_mapping(
+            old_exploration.id, old_exploration.version)
+        new_state_id_mapping = (
+            old_state_id_mapping.create_mapping_for_new_version(
+                old_exploration, exploration, change_list))
+    elif exploration.version == 1:
+        # Get state id mapping for first version of exploration.
+        new_state_id_mapping = (
+            exp_domain.StateIDMapping.create_mapping_for_new_exploration(
+                exploration))
+    else:
```

Comment:
Actually, here, you can do:

    if exploration.version == 1:
        ...
    else:
        ...

I don't think any other case will occur....? And I switched them because it's cleaner to do the special case first.

Reply:
Done